crescent/internal/handle_resp.py

```diff
@@ -72,21 +72,30 @@ async def handle_resp(event: InteractionCreateEvent) -> None:
     await _handle_slash_resp(bot, command, ctx)
 
 
+async def _handle_hooks(hooks: Sequence[HookCallbackT], ctx: Context) -> bool:
+    """Returns `False` if the command should not be run."""
+    for hook in hooks:
+        hook_res = await hook(ctx)
+
+        if hook_res and hook_res.exit:
+            return False
+    return True
+
+
 async def _handle_slash_resp(
     bot: Bot, command: MetaStruct[CommandCallbackT, AppCommandMeta], ctx: Context
 ) -> None:
-    for hook in command.metadata.hooks:
-        hook_res = await hook(ctx)
 
-        if hook_res and hook_res.exit:
-            break
+    if not await _handle_hooks(command.metadata.hooks, ctx):
+        return
 
-    else:
-        try:
-            await command.callback(ctx, **ctx.options)
-        except Exception as exc:
-            handled = await command.app._command_error_handler.try_handle(exc, [exc, ctx])
-            await bot.on_crescent_command_error(exc, ctx, handled)
+    try:
+        await command.callback(ctx, **ctx.options)
+    except Exception as exc:
+        handled = await command.app._command_error_handler.try_handle(exc, [exc, ctx])
+        await bot.on_crescent_command_error(exc, ctx, handled)
+
+    await _handle_hooks(command.metadata.after_hooks, ctx)
```

Comment:
Should these run even if the command errored out? Something to think about

Also this opens up some opportunities which will disappear once we merge (would be breaking changes). We could let command callbacks return a response that is passed to after hooks. Seems like stuff that complex should just be a function call...

In any case, if these do run even if the command raised an exception, we could add an `error` parameter to hooks, which is always None unless the hook is (1) an after_hook and (2) the command actually raised an exception.

We can also make this `error` parameter optional (by parsing the co_varnames of the hook and checking if it exists), so that the original syntax can be used.

Reply:
I think they should not run on an error. The exception cleanup system is `crescent.catch`. I dont think hooks should serve the same purpose. (I imagine after hooks are more for tracking data and posting it to a database, like my guessing game bot that had many commands that tracked how many guesses you did and score)

Reply:
Fair enough